File: engine/dynamic/render_vulkan/src/renderer/2d/scene_compiler.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/vector.hpp"

#include "argus/render/util/object_processor.hpp"
#include "argus/render/2d/scene_2d.hpp"

#include "internal/render_vulkan/renderer/shader_mgmt.hpp"
#include "internal/render_vulkan/renderer/2d/object_proc_impl.hpp"
#include "internal/render_vulkan/renderer/2d/scene_compiler.hpp"
#include "internal/render_vulkan/state/processed_render_object.hpp"
#include "internal/render_vulkan/state/render_bucket.hpp"
#include "internal/render_vulkan/state/scene_state.hpp"

namespace argus {
    static BucketKey _get_bucket_key(ProcessedRenderObject &processed_obj) {
        return BucketKey { processed_obj.material_res.uid, processed_obj.atlas_stride, processed_obj.z_index };
    }

    static void _handle_new_obj(Scene2DState &scene_state, ProcessedRenderObject &processed_obj) {
        RenderBucket *bucket;
        auto key = _get_bucket_key(processed_obj);
        auto existing_bucket_it = scene_state.render_buckets.find(key);
        if (existing_bucket_it != scene_state.render_buckets.end()) {
            bucket = existing_bucket_it->second;
        } else {
            bucket = &RenderBucket::create(processed_obj.material_res, processed_obj.atlas_stride,
                    processed_obj.z_index);
            scene_state.render_buckets[key] = bucket;
        }

        bucket->objects.push_back(&processed_obj);
        bucket->needs_rebuild = true;

        processed_obj.newly_created = false;
    }

    static void _handle_stale_obj(Scene2DState &scene_state, ProcessedRenderObject &processed_obj) {
        auto &state = scene_state.parent_state;

        deinit_object_2d(state, processed_obj);

        // we need to remove it from its containing bucket and flag the bucket for a rebuild
        auto *bucket = scene_state.render_buckets[_get_bucket_key(processed_obj)];
        remove_from_vector(bucket->objects, &processed_obj);
        bucket->needs_rebuild = true;

        processed_obj.destroy();
    }

    void compile_scene_2d(const Scene2D &scene, Scene2DState &scene_state) {
        process_objects_2d(scene, scene_state.processed_objs, create_processed_object_2d, update_processed_object_2d,
                &scene_state);

        for (auto it = scene_state.processed_objs.begin(); it != scene_state.processed_objs.end();) {
            auto &processed_obj = *reinterpret_cast<ProcessedRenderObject *>(it->second);
            if (processed_obj.newly_created) {
                _handle_new_obj(scene_state, processed_obj);
            } else if (!processed_obj.visited) {
                // wasn't visited this iteration, must not be present in the scene graph anymore
                _handle_stale_obj(scene_state, processed_obj);
                it = scene_state.processed_objs.erase(it);
                continue;
            }

            processed_obj.visited = false;
            ++it;
        }
    }
}

File: engine/modules/render_vulkan/src/renderer/scene_compiler.rs
use std::sync::atomic::Ordering;
use argus_render::util::process_objects_2d;
use vk_wrapper::vk::MAX_FRAMES_IN_FLIGHT;
use crate::renderer::object_proc_impl::process_object;
use crate::state::{ProcessedObject, RenderBucket, RenderBucketKey, RendererState};

fn get_bucket_key(processed_obj: &ProcessedObject) -> RenderBucketKey {
    RenderBucketKey::new(
        processed_obj.material_res.get_prototype().uid.clone(),
        &processed_obj.atlas_stride,
        processed_obj.z_index,
        processed_obj.light_opacity,
    )
}

pub(crate) fn compile_scene_2d(
    state: &mut RendererState<'_>,
    scene_id: &str,
) {
    process_objects_2d(
        scene_id,
        process_object,
        state,
    );
    
    let scene_state = state.scene_states_2d.get_mut(scene_id).unwrap();

    // objects not visited during the last iteration
    // must not be present in the scene graph anymore
    let stale_objs = scene_state.processed_objs
        .extract_if(|_, obj| !obj.newly_created && !obj.visited)
        .collect::<Vec<_>>();
    if !stale_objs.is_empty() {
        let last_frame = (state.cur_frame.load(Ordering::Relaxed) + MAX_FRAMES_IN_FLIGHT - 1) %
            MAX_FRAMES_IN_FLIGHT;
        let last_frame_fence = &state.swapchain.as_ref().unwrap().in_flight_fence[last_frame];
        state.device.wait_for_fences(&[last_frame_fence], true, u64::MAX).unwrap();

        for (stale_obj_handle, stale_obj) in stale_objs {
            let bucket = scene_state.render_buckets.get_mut(&get_bucket_key(&stale_obj)).unwrap();
            bucket.objects.retain(|h| h != &stale_obj_handle);
            bucket.needs_rebuild = true;
            stale_obj.destroy();
        }
    }

    for (obj_handle, obj) in &mut scene_state.processed_objs {
        if obj.newly_created {
            let key = get_bucket_key(obj);
            let bucket = scene_state.render_buckets.entry(key).or_insert_with(|| {
                RenderBucket::new(
                    obj.material_res.clone(),
                    obj.atlas_stride,
                    obj.z_index,
                    obj.light_opacity,
                )
            });

            bucket.objects.push(*obj_handle);
            bucket.needs_rebuild = true;

            obj.newly_created = false;
        }

        // reset visited flag for next iteration
        obj.visited = false;
    }
}
